sdks/python/apache_beam/runners/interactive/recording_manager.py

```diff
@@ -0,0 +1,334 @@
+    ]
+    reader, _ = cache_manager.read('full', self._cache_key,
+                                   limiters=limiters,
+                                   tail=tail)
+
+    # Because a single TestStreamFileRecord can yield multiple elements, we
+    # limit the count again here in the to_element_list call.
+    #
+    # There are two ways of exiting this loop either a limiter was triggered or
+    # all elements from the cache were read. In the latter situation, it may be
+    # the case that the pipeline was still running. Thus, another invocation of
+    # `read` will yield new elements.
+    for e in utils.to_element_list(reader,
+                                   coder,
+                                   include_window_info=True,
+                                   n=self._n):
+      yield e
+
+    # A limiter being triggered means that we have fulfilled the user's request.
+    # This implies that reading from the cache again won't yield any new
+    # elements. WLOG, this applies to the user pipeline being terminated.
+    if any(l.is_triggered()
+           for l in limiters) or ie.current_env().is_terminated(self._pipeline):
+      self._done = True
+
+
+class Recording:
+  """A group of PCollections from a given pipeline run."""
+  def __init__(
+      self,
+      user_pipeline,  # type: beam.Pipeline
+      pcolls,  # type: List[beam.pvalue.PCollection]
+      result,  # type: beam.runner.PipelineResult
+      pipeline_instrument,  # type: beam.runners.interactive.PipelineInstrument
+      max_n,  # type: int
+      max_duration_secs  # type: float
+      ):
+
+    self._user_pipeline = user_pipeline
+    self._result = result
+    self._pcolls = pcolls
+
+    pcoll_var = lambda pcoll: pipeline_instrument.cacheable_var_by_pcoll_id(
+        pipeline_instrument.pcolls_to_pcoll_id.get(str(pcoll), None))
+
+    self._streams = {
+        pcoll: ElementStream(
+            pcoll,
+            pcoll_var(pcoll),
+            pipeline_instrument.cache_key(pcoll),
+            max_n,
+            max_duration_secs)
+        for pcoll in pcolls
+    }
+    self._start = time.time()
+    self._duration_secs = max_duration_secs
+    self._set_computed = bcj.is_cache_complete(str(id(user_pipeline)))
+
+    # Run a separate thread for marking the PCollections done. This is because
+    # the pipeline run may be asynchronous.
+    self._mark_computed = threading.Thread(target=self._mark_all_computed)
+    self._mark_computed.daemon = True
+    self._mark_computed.start()
+
+  def _mark_all_computed(self):
+    # type: () -> None
+
+    """Marks all the PCollections upon a successful pipeline run."""
+    if not self._result:
+      return
+
+    while not PipelineState.is_terminal(self._result.state):
+      if time.time() - self._start >= self._duration_secs:
+        self._result.cancel()
+        self._result.wait_until_finish()
+
+      if all(s.is_done() for s in self._streams.values()):
```

Comment:
Should this be an elif, since you shouldn't need to cancel twice?

Reply:
Cancel is an idempotent operation so I didn't really think about it. I can change it to elif though